plugin-server/src/worker/ingestion/event-pipeline/cookielessServerHashStep.ts

```diff
@@ -0,0 +1,303 @@
+
+// Due to ingestion lag and time zones, we allow creating the hash for a calendar day that is different to the current
+// UTC calendar day, provided somewhere in the world could be in that calendar day, with some additional buffer in the
+// past for ingestion lag.
+
+// To ensure that a user that logs in and out doesn't collide with themselves, we store the set of identify event UUIDs
+// in redis against the base hash value, and use the number of identifies for that hash to calculate the final hash value.
+// The exact number we append is the number of identifies that happened before this user called identify, which will
+// mean adjusting the number based on whether the event was pre or post identify, or was itself the identify event.
+
+const TIMEZONE_FALLBACK = 'UTC'
+const SENTINEL_COOKIELESS_SERVER_HASH_DISTINCT_ID = '$sentinel_cookieless_server_hash'
+const MAX_NEGATIVE_TIMEZONE_HOURS = 12
+const MAX_POSITIVE_TIMEZONE_HOURS = 14
+const MAX_INGESTION_LAG_HOURS = 24
+const SALT_TTL_SECONDS =
+    (MAX_POSITIVE_TIMEZONE_HOURS + MAX_NEGATIVE_TIMEZONE_HOURS + MAX_INGESTION_LAG_HOURS + 24) * 60 * 60
+const SESSION_TTL_SECONDS = 60 * 60 * 24
+
+export async function cookielessServerHashStep(
+    runner: EventPipelineRunner,
+    event: PluginEvent
+): Promise<[PluginEvent | undefined]> {
+    // if events aren't using this mode, skip all processing
+    if (event.properties?.['$device_id'] !== SENTINEL_COOKIELESS_SERVER_HASH_DISTINCT_ID) {
+        return [event]
+    }
+
+    // if the team isn't allowed to use this mode, drop the event
+    const team = await runner.hub.teamManager.getTeamForEvent(event)
+    if (!team?.cookieless_server_hash_opt_in) {
+        // TODO log
+        return [undefined]
+    }
+
+    const timestamp = event.timestamp ?? event.sent_at ?? event.now
+
+    // drop some events that aren't valid in this mode
+    if (!timestamp) {
+        // TODO log
+        return [undefined]
+    }
+    const sessionId = event.properties['$session_id']
+    if (sessionId !== SENTINEL_COOKIELESS_SERVER_HASH_DISTINCT_ID) {
+        // TODO log
+        return [undefined]
+    }
+    // if it's an identify event, it must have the sentinel distinct id
+    if (
+        event.event === '$identify' &&
+        event.properties['$anon_distinct_id'] !== SENTINEL_COOKIELESS_SERVER_HASH_DISTINCT_ID
+    ) {
+        // TODO log
+        return [undefined]
+    }
+
+    const { userAgent, ip, host, timezone, timestampMs, teamId } = getProperties(event, timestamp)
+    if (!userAgent || !ip || !host) {
+        // TODO log
+        return [undefined]
+    }
+
+    // TRICKY: if a user were to log in and out, to avoid collisions, we would want a different hash value, so we store the set of identify event uuids for identifies
+    // ASSUMPTION: all events are processed in order, for this to happen we need them to be in the same kafka topic at this point
```

Comment:
What if someone is calling `$identify` much more often than they need? For example, in the backend? Are we properly handling this?

Also, newbie question, why would someone want cookieless tracking if they've identified the person?

Reply:
We want users to be upgradeable from cookieless and anon -> identified, meaning that you'll get the best of both worlds tracking-wise (users who don't give consent can still be tracked somewhat, and users who do give consent are tracked really well)

Reply:
When this upgrade happens, we want the events to be part of the same session before and after, so we need to continue using the session logic below